### PR TITLE
[riot-patch-part-1] Integrate standalone build support

### PR DESCRIPTION
**Pull request by j-devel**
This is "part-1" of [the RIOT summit's follow-up efforts](https://forum.riot-os.org/t/micropython/478) to get RIOT-OS to be integrated into upstream micropython as one of [the official ports](https://github.com/micropython/micropython/tree/master/ports).

As a base, we use the commits derived from the "add_riot_port" branch of https://github.com/kaspar030/micropython.  On top of that, in 8f6bda0, we add

1) build compatibility fixes for micropython v1.17 ('ports/riot/{Makefile,help.c,mp_riot.c}'),
2) adaptation of [the micropython pkg](https://github.com/RIOT-OS/RIOT/tree/423da42d1cebbb60a6fb15cd13a211ce7e7e405a/pkg/micropython) currently in RIOT (as 'ports/riot/{Makefile.dep,Makefile.include,micropython.h}'),
3) adaptation of [the micropython example](https://github.com/RIOT-OS/RIOT/tree/423da42d1cebbb60a6fb15cd13a211ce7e7e405a/examples/micropython) currently in RIOT (as 'ports/riot/main/*'), and
4) a custom .mk file for evaluating this PR ('ports/riot/summit.mk').

Through 2) and 3), we intend to combine the currently scattered RIOT-specific micropython resources into the standalone port directory (i.e. 'ports/riot') that has 'main.c' therein.  (I think this treatment better conforms to the rules of [other existing ports](https://github.com/micropython/micropython/tree/master/ports).)

## Known issues/limitations

- [x] [1/2 of [the achievements](https://github.com/RIOT-OS/micropython/pull/5#issuecomment-951752019)] make it possible to build micropython for RIOT together with RIOT, initiated from a micropython checkout
- [ ] [2/2 of the achievements] have micropython as a package in RIOT, so RIOT applications can pull it in?
- [x] ~~[When using RIOT of latest master, build fails with `ld` error](https://github.com/RIOT-OS/micropython/pull/5/files#r725636594)~~ (resolved by 3e378fe81)
- [ ] How to enforce outside-source-tree build for micropython? (and provably a corollary to fix: [genhdr/qstrdefs.generated.h workaround](https://github.com/RIOT-OS/micropython/pull/5/files#r725634854))
- [x] ~~A wrong micropython version (e.g. v1.11-*) is being reported in REPL~~ (just make sure all git tags are checked out in the working branch)

## Testing

To build MCU "native", `cd` to the 'ports/riot' directory and invoke `make`:

```
$ cd micropython/ports/riot
$ make
(snip)
Cloning into 'RIOT'...
(snip)
Building application "main" for "native" with MCU "native".
(snip)
   text	   data	    bss	    dec	    hex	filename
 279336	  12872	 199960	 492168	  78288	<path/to>/micropython/ports/riot/main/bin/native/main.elf
$ ./main/bin/native/main.elf
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2022.01-devel-280-g6029f2)
sizeof(mp_heap): 131072
-- Executing boot.py
boot.py: MicroPython says hello!
-- boot.py exited. Starting REPL..
MicroPython v1.17-26-gf9787e9cd on 2021-10-28; riot-native with native
Type "help()" for more information.
>>>
```

### CODE REVIEW

File: ports/riot/summit.mk [hunk 1]
```diff
@@ -0,0 +1,15 @@
+all: build
+
+GENHDR_DIR := ../../genhdr
+
+RIOT:
+	git -c advice.detachedHead=false \
+		clone --branch 2021.04 https://github.com/RIOT-OS/RIOT
+
+build: RIOT
+	mkdir -p $(GENHDR_DIR) && touch $(GENHDR_DIR)/qstrdefs.generated.h
```
Comment (j-devel) [inline ports/riot/summit.mk]: FIXME: Build breaks without this workaround

File: ports/riot/summit.mk [hunk 2]
```diff
@@ -0,0 +1,15 @@
+all: build
+
+GENHDR_DIR := ../../genhdr
+
+RIOT:
+	git -c advice.detachedHead=false \
+		clone --branch 2021.04 https://github.com/RIOT-OS/RIOT
```
Comment (j-devel) [inline ports/riot/summit.mk]: FIXME: when using RIOT of latest master, build fails with `collect2: error: ld returned 1 exit status`

Comment (j-devel) [reply ports/riot/summit.mk]: Resolved by 3e378fe81

**Comment by kaspar030** [comment]
I can build this using summit.mk.
I'm not entirely happy that the use of EXTERNAL_MODULE_DIRS is forcing the micropython module name to `riot.a`, that will probably bite us ... Like, `USEMODULE += riot` to select micropython is at least confusing. I tried just re-creating this using DIRS, but the machinery is not nice for that.

Just to be clear, we want to:
1. make it possible to build micropython for RIOT together with RIOT, initiated from a micropython checkout (like, as the supplied summit.mk is doing)
2. have micropython as a package in RIOT, so RIOT applications can pull it in?

Maybe we can re-use RIOT's package machinery for supporting both configurations. By default, RIOT expects some remoge (e.g., git) package source. It allows overriding the remote source using a local path. We could maybe use a flag (BUILD_FROM_MICROPYTHON) that overrides this, if initiating the build from within micropython.

Let me try that ...

**Comment by kaspar030** [comment]
> Let me try that ...

hm, this would work, *if* all Makefile.dep and Makefile.include info is available in the RIOT tree, which would cause synchronization issues.

**Comment by kaspar030** [comment]
@j-devel if we move the actual micropython module a bit, we can get the "micropython" name back.

I did that here: https://github.com/kaspar030/micropython/commit/1372f20f06cf18d538dc98844253a0b372af91ce

What do you think?

**Comment by kaspar030** [comment]
Otherwise I suggest integrating summit.mk somehow into Makefile, guarded with a flag? so anyone typing "make" in `ports/riot` builds something. `modules/micropython/Makefile` could set that flag when calling the `ports/riot` make.

**Comment by j-devel** [comment]
Thanks @kaspar030 for the review and advice!

> if we move the actual micropython module a bit, we can get the "micropython" name back. I did that here: kaspar030@1372f20

> I suggest integrating summit.mk somehow into Makefile

I think your micropython module name fix is really nice (where 'modules/micropython/Makefile' also gives us a chance to introduce a flag for refactoring 'summit.mk' as you say).  I have pulled the commit and managed to integrate 'summit.mk' into 'Makefile'.  f9787e9

**Comment by j-devel** [comment]
> Just to be clear, we want to:
>
> 1. make it possible to build micropython for RIOT together with RIOT, initiated from a micropython checkout
> 2. have micropython as a package in RIOT, so RIOT applications can pull it in?

I have included the status of these items in the task list.

For 2), I would like to do some new experiments to make sure that RIOT applications can really consume this micropython repo as part of a RIOT package.

**Comment by kaspar030** [comment]
I think we can go with this. Are you happy with the commits and their naming? do we want the workflows disabled (do we mind if they fail)?

**Comment by kaspar030** [comment]
> I think we can go with this. Are you happy with the commits and their naming? do we want the workflows disabled (do we mind if they fail)?

what I mean is, let's merge this and tackle 2) as next step. once that is solved, we can maybe already change the RIOT package to point to this repo.

**Comment by j-devel** [comment]
> I think we can go with this. Are you happy with the commits and their naming?

Yes, I am.

> do we want the workflows disabled (do we mind if they fail)?

I think we can simply keep going without minding the workflows failure; so reverted in 3d54d2b

> what I mean is, let's merge this and tackle 2) as next step. once that is solved, we can maybe already change the RIOT package to point to this repo.

Exactly.

**Review by kaspar030 — APPROVED** [review]
ACK.

**Pull request merged**